src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go

```diff
@@ -386,6 +386,19 @@ func (s *cloudService) StartVM(ctx context.Context, req *pb.StartVMRequest) (res
 	}
 
 	instance, err := s.provider.CreateInstance(ctx, sandbox.podName, string(sid), sandbox.cloudConfig, sandbox.spec)
+
+	// Cleanup instance if it was created but an error occurred (either during creation or later)
+	defer func() {
+		if err != nil && instance != nil && instance.ID != "" {
+			logger.Printf("cleaning up instance %s due to error: %v", instance.ID, err)
+			cleanupCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```

Comment:
This is an essential fix, thanks! LGTM overall. 
nit, IDK if the context timeout is always honored, however,  I'd use even larger timeout, i recall seeing longer deletion time